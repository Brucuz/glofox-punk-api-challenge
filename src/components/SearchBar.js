import React from "react";
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import { Textbox } from "react-inputs-validation";
import "react-inputs-validation/lib/react-inputs-validation.min.css";
import { fetchBeers } from "../actions";
import { connect } from "react-redux";

const TEXT_REGEX = /^[0-9A-Za-z\s\-]+$/;
class SearchBar extends React.Component {
  constructor(props) {
    super(props);
    this.searchOptions = [
      { value: "name", label: "Name" },
      { value: "brewed_before", label: "Brewed before" }
    ];
    this.state = {
      searchBy: "name",
      searchParams: { name: "" }
    };
  }

  changeSearchParams(searchBy) {
    this.setState({
      searchBy,
      searchParams:
        searchBy === "name" ? { beer_name: "" } : { brewed_before: new Date() }
    });
  }

  selectDate(date) {
    this.setState({
      searchParams: { brewed_before: date }
    });
  }

  searchBeers(event) {
    event.preventDefault();
    if(this.state.searchBy === 'date' || TEXT_REGEX.test(this.state.searchParams.beer_name)){
      this.props.fetchBeers(this.state.searchParams);
    }
  }
  renderInputParam() {
    if (this.state.searchBy === "name") {
      return (
        <Textbox
          attributesInput={{
            type: "text",
            placeholder: "Search for some beers!"
          }}
          value={this.state.searchParams.beer_name}
          onChange={(name, e) => {
            this.setState({ searchParams: { beer_name:name } });
          }}
          onBlur={e => {}}
          validationOption={{
            reg: TEXT_REGEX,
            regMsg: "Use only letters, numbers, hyphens and spaces."
          }}
        />
      );
    }
    return (
      <DatePicker
        selected={this.state.searchParams.brewed_before}
        onChange={date => this.selectDate(date)}
      />
    );
  }

  renderSearchOptions() {
    return this.searchOptions.map(option => {
      return (
        <div className="form-check" key={option.value}>
          <input
            className="form-check-input"
            type="radio"
            style={{ display: "block", opacity: 1 }}
            value={option.value}
            checked={this.state.searchBy === option.value}
            onChange={() => this.changeSearchParams(option.value)}
          />
          <label className="form-check-label">{option.label}</label>
        </div>
      );
    });
  }

  render() {
    return (
      <div className="SearchBar">
        <form className="row" onSubmit={event => this.searchBeers(event)}>
          <div className="form-group col-md-3 col-xs-12">
            {this.renderInputParam()}
          </div>
          <div className="col-md-6 col-xs-12">{this.renderSearchOptions()}</div>
          <div className="form-group col-md-3 col-xs-12">
            <button type="submit" className="btn btn-primary">
              Submit
            </button>
          </div>
        </form>
      </div>
    );
  }
}


export default connect(null, { fetchBeers })(
  SearchBar
);
