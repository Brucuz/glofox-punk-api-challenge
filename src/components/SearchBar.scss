.SearchBar {
  padding-top: 20px;
  &-button {
    display: flex;
    align-items: center;
    justify-content: center;
    button {
      width: 80%;
    }
  }
  .search-by {
    display: flex;
    .form-check {
      margin: 0 15px;
    }
  }
  .react-inputs-validation__msg_identifier {
    position: absolute;
    top: calc(100% - 20px);
  }
  input[type="text"].react-inputs-validation__textbox__input___20hDL {
    margin-bottom: 20px;
  }
  .react-datepicker-wrapper {
    input {
      border: 1px solid #e5e5e5;
    }
  }
  @media (max-width: 576px) {
    &-button {
      padding: 20px 0;
      button{
          width: 76%;
      }
    }
    input[type="text"].react-inputs-validation__textbox__input___20hDL {
      width: 80%;
      left: 10%;
      position: relative;
    }
    .react-inputs-validation__msg_identifier {
      left: 12%;
    }
    .search-by {
      justify-content: center;
    }
  }
}
