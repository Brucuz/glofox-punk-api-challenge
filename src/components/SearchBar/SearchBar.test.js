import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import SearchBar from "./SearchBar";

const mockStore = configureStore([]);

let store;
let component;

beforeEach(() => {
  store = mockStore({
    beerList: []
  });

  store.dispatch = jest.fn();

  component = renderer.create(
    <Provider store={store}>
      <SearchBar />
    </Provider>
  );
});

it("renders correctly", () => {
  expect(component).toMatchSnapshot();
});

it("dispatch action search_beer", async () => {
  renderer.act(() => {
    component.root.findByType("form").props.onSubmit(new Event("submit"));
  });

  expect(store.dispatch).toHaveBeenCalledTimes(1);
});
