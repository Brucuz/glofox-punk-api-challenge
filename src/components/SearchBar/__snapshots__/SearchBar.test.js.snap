// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="SearchBar"
>
  <form
    className="row"
    onSubmit={[Function]}
  >
    <div
      className="form-group col-sm-3 col-xs-12"
    >
      <div
        className=" react-inputs-validation__textbox__wrapper___3tnXv false false false"
        style={Object {}}
      >
        <div
          className=" undefined false false false"
          style={Object {}}
        >
          <input
            className=" react-inputs-validation__textbox__input___20hDL false false false"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            placeholder="Search for some beers!"
            style={Object {}}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="search-by col-sm-6 col-xs-12"
    >
      <div
        className="form-check"
      >
        <input
          checked={true}
          className="form-check-input"
          onChange={[Function]}
          style={
            Object {
              "display": "block",
              "opacity": 1,
            }
          }
          type="radio"
          value="name"
        />
        <label
          className="form-check-label"
        >
          Name
        </label>
      </div>
      <div
        className="form-check"
      >
        <input
          checked={false}
          className="form-check-input"
          onChange={[Function]}
          style={
            Object {
              "display": "block",
              "opacity": 1,
            }
          }
          type="radio"
          value="brewed_before"
        />
        <label
          className="form-check-label"
        >
          By brewed before
        </label>
      </div>
    </div>
    <div
      className="SearchBar-button form-group col-sm-3 col-xs-12"
    >
      <button
        className="btn btn-primary"
        id="search-button"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
</div>
`;
